### ISSUE TITLE
Issue #6: Implementera ML-baserad Smart Kategorisering

### ISSUE THREAD

**Issue by pownas** [pownas/Privatekonomi#174]
**Labels:** `feature`, `ml`, `transactions`, `ai`, `high-priority`  
**Assignees:** (lägg till efter behov)  
**Milestone:** Fas 1 - Kritiska Förbättringar  
**Estimat:** 7-10 dagar

### Beskrivning

Förbättra automatisk kategorisering med maskininlärning som lär sig från användarens egna kategoriseringsmönster.

### Bakgrund

- ✅ Regelbaserad kategorisering finns redan
- ⚠️ 44+ förkonfigurerade regler, men inte personliga
- 🎯 ML kan lära sig användarens unika mönster
- 🎯 Förbättras automatiskt över tid

### Problem som Löses

1. **Regelbaserad begränsning**: Nuvarande system kan inte lära sig nya mönster
2. **Generiska regler**: Samma regler för alla användare
3. **Manuell uppdatering**: Användare måste manuellt skapa nya regler
4. **Låg precision**: Svårt att hantera edge cases

### Funktionalitet

**Core Features:**
- [x] Träna ML-modell på användarens egna data
- [x] Lär sig från användarens manuella kategoriseringar
- [x] Föreslå kategorier med konfidenspoäng (0-100%)
- [x] Markera som "osäker" om konfidensen är låg (<70%)
- [x] Kontinuerlig förbättring genom feedback-loop
- [x] Export av träningsdata för analys
- [x] Fallback till regelbaserad om ML ej tillämplig

**Användningsfall:**
```
Scenario 1: Ny transaktion från okänd källa
Input: "Pizzeria Milano - 285 kr"
ML Output:
  - Kategori: Mat & Dryck (Restaurang)
  - Konfidenspoäng: 92%
  - Förslag: ✓ Acceptera automatiskt

Scenario 2: Osäker transaktion
Input: "Gekås Ullared - 1,250 kr"
ML Output:
  - Kategori: Shopping (45%), Boende (30%), Mat (25%)
  - Konfidenspoäng: 45% (OSÄKER)
  - Förslag: ? Fråga användare
```

### Teknisk Implementation

**ML Framework:**
- **ML.NET** - Microsoft's ML framework för .NET
  - URL: https://dotnet.microsoft.com/apps/machinelearning-ai/ml-dotnet
  - Stöd för: Classification, Regression, Clustering
  - Integreras smidigt med befintlig .NET kod

**Algoritm:**
Föreslagna algoritmer i prioritetsordning:
1. **SDCA (Stochastic Dual Coordinate Ascent)** - Snabb multi-class classification
2. **LightGBM** - Gradient boosting för högre precision
3. **Naive Bayes** - Baseline, enkel och snabb

**Features för modellen:**
```csharp
public class TransactionFeatures
{
    // Text features (TF-IDF)
    [VectorType(1000)]
    public float[] DescriptionTfIdf { get; set; }
    
    // Numeric features
    public float Amount { get; set; }
    public float AmountLog { get; set; } // Log-transform för bättre distribution
    public float DayOfWeek { get; set; }
    public float DayOfMonth { get; set; }
    public float Hour { get; set; } // Om tidstämpel finns
    
    // Categorical features (one-hot encoded)
    public bool IsWeekend { get; set; }
    public bool IsMonthStart { get; set; }
    public bool IsMonthEnd { get; set; }
    
    // Label (target)
    [ColumnName("Label")]
    public string Category { get; set; }
}
```

**Model Training Pipeline:**
```csharp
public interface ITransactionMLService
{
    // Training
    Task TrainModelAsync(string userId);
    Task<ModelMetrics> EvaluateModelAsync(string userId);
    
    // Prediction
    Task<CategoryPrediction> PredictCategoryAsync(Transaction transaction);
    Task<List<CategoryPrediction>> PredictBatchAsync(List<Transaction> transactions);
    
    // Model management
    Task SaveModelAsync(string userId, string modelPath);
    Task LoadModelAsync(string userId);
    Task<bool> IsModelTrainedAsync(string userId);
    
    // Feedback loop
    Task UpdateModelWithFeedbackAsync(string userId, Transaction transaction, string correctCategory);
}

public class CategoryPrediction
{
    public string Category { get; set; }
    public float Confidence { get; set; } // 0-1
    public bool IsUncertain => Confidence < 0.7f;
    public Dictionary<string, float> AlternativeCategories { get; set; } // Top 3
}
```

**Datamodell:**
```csharp
// MLModel.cs
public class MLModel
{
    public int ModelId { get; set; }
    public string UserId { get; set; }
    public string ModelPath { get; set; } // Path to saved model file
    public DateTime TrainedAt { get; set; }
    public int TrainingRecordsCount { get; set; }
    public float Accuracy { get; set; }
    public float Precision { get; set; }
    public float Recall { get; set; }
    public string Metrics { get; set; } // JSON med detaljerad metrics
}

// UserFeedback.cs
public class UserFeedback
{
    public int FeedbackId { get; set; }
    public string UserId { get; set; }
    public int TransactionId { get; set; }
    public string PredictedCategory { get; set; }
    public float PredictedConfidence { get; set; }
    public string ActualCategory { get; set; } // Vad användaren valde
    public bool WasCorrectionNeeded { get; set; }
    public DateTime FeedbackDate { get; set; }
}
```

### Åtgärder

**Fas 1: ML Infrastructure (3 dagar)**
- [ ] Installera ML.NET NuGet package
- [ ] Skapa `ITransactionMLService` interface
- [ ] Implementera `TransactionMLService`
- [ ] Skapa modeller: `MLModel`, `UserFeedback`
- [ ] Setup model storage (lokal fil eller Azure Blob)

**Fas 2: Feature Engineering (2 dagar)**
- [ ] Implementera text preprocessing (TF-IDF)
- [ ] Extrahera numeric features från transaktioner
- [ ] Skapa feature transformation pipeline
- [ ] Testa feature quality

**Fas 3: Model Training (2 dagar)**
- [ ] Implementera training pipeline
- [ ] Cross-validation (80/20 split)
- [ ] Hyperparameter tuning
- [ ] Model evaluation metrics
- [ ] Save/Load model functionality

**Fas 4: Prediction Integration (2 dagar)**
- [ ] Integrera ML-predictions i TransactionService
- [ ] Fallback-logik till regelbaserad
- [ ] Confidence threshold handling
- [ ] Batch prediction för import

**Fas 5: UI & Feedback Loop (2 dagar)**
- [ ] Visa ML-förslag i UI med konfidenspoäng
- [ ] "Osäker"-indikator för låg confidence
- [ ] Feedback-knapp ("Rätt kategori", "Fel kategori")
- [ ] Auto-reträning varje natt (background service)

**Fas 6: Monitoring & Analytics (1 dag)**
- [ ] Model metrics dashboard
- [ ] Accuracy över tid
- [ ] Användnings-statistik
- [ ] Export träningsdata för analys

### Berörd Kod

- `src/Privatekonomi.Core/ML/TransactionMLService.cs` (ny)
- `src/Privatekonomi.Core/ML/ITransactionMLService.cs` (ny)
- `src/Privatekonomi.Core/Models/MLModel.cs` (ny)
- `src/Privatekonomi.Core/Models/UserFeedback.cs` (ny)
- `src/Privatekonomi.Core/Services/TransactionService.cs` (uppdatera)
- `src/Privatekonomi.Web/Components/Pages/Transactions.razor` (uppdatera)
- `src/Privatekonomi.Web/Components/Dialogs/EditTransactionDialog.razor` (uppdatera)

### Training Requirements

**Minimum krav:**
- Minst 50 kategoriserade transaktioner per användare
- Minst 5 exempel per kategori
- Om ej tillräcklig data: fallback till regelbaserad

**Optimal:**
- 200+ transaktioner för bättre precision
- Jämn fördelning över kategorier
- Historik över minst 3 månader

### Testning

- [ ] Enhetstester för `TransactionMLService`
- [ ] Testa med små dataset (<50 trans)
- [ ] Testa med stora dataset (>1000 trans)
- [ ] Test accuracy vs regelbaserad
- [ ] Performance-test (prediction latency)
- [ ] Test feedback loop
- [ ] Cross-user test (modeller ska inte läcka)

### Metrics & Success Criteria

**Target Metrics:**
- Accuracy: >85% (vs regelbaserad ~75%)
- Precision: >80%
- Recall: >80%
- Prediction latency: <100ms per transaktion
- Confidence calibration: 90% av "högkonfidens" ska vara korrekta

### Säkerhet & Privacy

- [ ] Modeller är user-specific (ingen delning mellan användare)
- [ ] Träningsdata lagras säkert
- [ ] GDPR: Rätt att radera modell och träningsdata
- [ ] Opt-in för ML (användare kan välja regelbaserad)

### Dokumentation

Efter implementation:
- [ ] README.md - Lägg till ML-kategorisering
- [ ] Skapa `docs/ML_CATEGORIZATION.md`
- [ ] API-dokumentation för `ITransactionMLService`
- [ ] Användarguide: "Hur ML-kategorisering fungerar"

### Framtida Förbättringar

- [ ] Deep Learning (LSTM/Transformer) för ännu bättre text-förståelse
- [ ] Transfer learning från andra användare (anonymiserat)
- [ ] Multi-label prediction (split categories)
- [ ] Anomali-detektion för ovanliga transaktioner
